### PR TITLE
Port to CMake build system

### LINKED ISSUE #1: CMake support
Add CMakeLists.txt

### PR DESCRIPTION
**Pull request by friendlyanon**
Generated a base project structure with [cmake-init](https://github.com/friendlyanon/cmake-init), which also adds CI.  
CI revealed that this library will not build on anything other than libstdc++.

I have left some things disabled, like coverage in CI, because that requires a codecov token.

I made the vendor and install interfaces as similar as possible, so people using something like CPM (or FetchContent directly) can also use this without trouble.

Read the HACKING document on how to develop with this PR.

Fixes #1

### CODE REVIEW

File: BUILDING.md
Comment (cdeln) [inline BUILDING.md:71]: Can you link to this document in the readme under "installation"?

File: Makefile
Comment (cdeln) [inline Makefile:73]: I am new to github CI. How does one reproduce what the CI does locally? Does it use Docker?
I actually can't build with magic enum support on my local machine, so I need to use a virtual env.

File: cmake/docs.cmake
Comment (cdeln) [inline cmake/docs.cmake:3]: Possible to require python version? I wonder if that is what does not work for me, I have 3.6.9 locally (ancient, I know...)

File: test/CMakeLists.txt
Comment (cdeln) [inline test/CMakeLists.txt:51]: It's astonishing how much code is required to do cmake compare to plain make... The price we pay for cross platform i guess :) Many thanks for doing this anyways, I really suck at cmake, nice to have it in place!
I would like to be able to build everything on my machine before merging this ofc.

File: .github/workflows/ci.yml
Comment (cdeln) [inline .github/workflows/ci.yml:20]: Cool, will do!

File: BUILDING.md
Comment (cdeln) [inline BUILDING.md:24]: Cool, I installed vcpkg.
Configured with

`cmake -S . -B build -D CMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE="$VCPKG_ROOT/scripts/buildsystems/vcpkg.cmake" -Denum_set_DEVELOPER_MODE=ON -DVCPKG_MANIFEST_FEATURES=magic-enum`

and try to build with

`cmake --build build`

but it can not find the magic enum header, do we need to set include path through cmake manually?

File: .github/workflows/ci.yml
Comment (cdeln) [inline .github/workflows/ci.yml:14]: Maybe set this to a fixed version and bump up manually to avoid spurious CI failures?

File: .github/workflows/ci.yml
Comment (cdeln) [inline .github/workflows/ci.yml:53]: Great! Can you expose a cmake target for doing the coverage report locally also?

File: HACKING.md
Comment (cdeln) [inline HACKING.md:98]: Link to this document in readme, maybe under the contribution section?
That section needs update anyways, now when we go from a pure Make workflow to CMake

File: CMakeLists.txt
Comment (cdeln) [inline CMakeLists.txt:84]: The `docs` target is broken for me locally, getting some weird python syntax errors.

See my comment below about Docker, I would like to be able to build it.

Would also be nice to have a documentation page served later!

File: CMakeLists.txt
Comment (cdeln) [inline CMakeLists.txt:67]: Why the `enum_set_` prefix for this option? BUILD_TESTING does not have it for example?

**Review by cdeln — CHANGES_REQUESTED** [review]
Many thanks for this changes, I really suck at cmake and have never touched github CI. This would have taken ages for me to set up. I have some comments which I would like to get fixed before merging, once I can run everything locally I am comfortable with merging this. Then at least, me and the CI starts of on good terms ;)

File: .github/workflows/ci.yml
Comment (cdeln) [reply .github/workflows/ci.yml:20]: I generated a token and make a secret called CODECOV_TOKEN.
That's it?

File: BUILDING.md
Comment (friendlyanon) [reply BUILDING.md:71]: I left the README untouched, because I don't how you'd prefer to structure it. If you grab cmake-init for yourself and look at what it generates by default, maybe that gives you some idea.

File: Makefile
Comment (friendlyanon) [reply Makefile:73]: You install vcpkg and then run the commands as is from either the CI workflow or the HACKING document. No virtual environment required.

File: cmake/docs.cmake
Comment (friendlyanon) [reply cmake/docs.cmake:3]: I don't know what's the minimum version [this script](https://github.com/mosra/m.css/blob/master/documentation/doxygen.py) supports.

File: test/CMakeLists.txt
Comment (friendlyanon) [reply test/CMakeLists.txt:51]: Check the HACKING guide. Following it should be enough to drop you into developer mode.

File: .github/workflows/ci.yml
Comment (friendlyanon) [reply .github/workflows/ci.yml:20]: Yes, it's nothing fancy. The comments there describe everything you have to do. I will not change this in my branch, because it's just going to fail anyway.

File: BUILDING.md
Comment (friendlyanon) [reply BUILDING.md:24]: The BUILDING document is for consumers of the library, not the developers. Everything is set up in terms of CMake.

File: .github/workflows/ci.yml
Comment (friendlyanon) [reply .github/workflows/ci.yml:14]: The version doesn't change often and even when it does change you are warned months before GitHub changes what the `latest` one points to. You can find the images and related news [here](https://github.com/actions/virtual-environments).

File: .github/workflows/ci.yml
Comment (friendlyanon) [reply .github/workflows/ci.yml:53]: I found that CTest's coverage support is not quite as good as the lcov generated document, so I left this as just a shell script. Since this is not cross-platform, I don't know what'd be the best way to have it in the CMake scripts. I have to think about this for cmake-init as well.

File: HACKING.md
Comment (friendlyanon) [reply HACKING.md:98]: Refer to my earlier reply regarding the README.

File: CMakeLists.txt
Comment (friendlyanon) [reply CMakeLists.txt:84]: I don't know what's the minimum Python version required by m.css.

Regarding the documentation page, you can have a GitHub CI workflow to automatically deploy the docs to a GitHub Pages branch.

File: CMakeLists.txt
Comment (friendlyanon) [reply CMakeLists.txt:67]: Variables that are not hidden behind the developer mode are prefixed with `enum_set_` to avoid name collision for consuming projects. Everything in the developer mode can have any name you want, because consumers will (should) never interact with them.

File: .github/workflows/ci.yml
Comment (cdeln) [reply .github/workflows/ci.yml:20]: Ok, I can turn it on later!

File: .github/workflows/ci.yml
Comment (cdeln) [reply .github/workflows/ci.yml:14]: Alright, I guess it is not too much to ask of a maintainer to keep an eye open for this every second year ;)

File: .github/workflows/ci.yml
Comment (cdeln) [reply .github/workflows/ci.yml:53]: I think you can implement any command in cmake using something like [add_custom_command](https://cmake.org/cmake/help/latest/command/add_custom_command.html) ? 
It doesnt have to be cross platform, we can mention it in the documentation that it is only works for ubuntu-latest.
I would like to be able to see coverage locally, since I had that before with gcov in the Makefile.
Would you like to take on the exercise? Then you can back-port the solution to cmake-init

File: BUILDING.md
Comment (cdeln) [reply BUILDING.md:71]: Feel free to update it, I think it should reflect the state of the repo as well as possible. Just a link, nothing more is needed.

File: cmake/docs.cmake
Comment (cdeln) [reply cmake/docs.cmake:3]: Alright, we can try to fix it later, I assume it works for you, and since it was not a feature before, let's bring it in and I can try to make it work for myself later.

File: HACKING.md
Comment (cdeln) [reply HACKING.md:98]: See my answer on that comment. Just a link in the contribution section is fine. You can remove the instructions for the Make-commands there since they are not supported anymore after merging this.

File: CMakeLists.txt
Comment (cdeln) [reply CMakeLists.txt:84]: Ok that sounds great! It will not do that after this changes right? Otherwise I will make an issue for it.

File: CMakeLists.txt
Comment (cdeln) [reply CMakeLists.txt:67]: Ah I understand!

File: Makefile
Comment (cdeln) [reply Makefile:73]: Hmm ok, I will try to follow the CI workflow step by step later today.
I think that this does not guarantee that it works though, magic_enum does not seem to work on ubuntu18.04 (with the default GCC toolchain), so I think i am pretty screwed there even with vcpkg no?
I can also update the Dockerfile to reflect the CI workflow, though it becomes two to things to maintain, which would be nice to avoid.

File: CMakeLists.txt
Comment (cdeln) [reply CMakeLists.txt:84]: I made an issue for it: https://github.com/cdeln/cpp_enum_set/issues/3

File: HACKING.md
Comment (cdeln) [reply HACKING.md:98]: I will do this in a later cleanup commit.

File: .github/workflows/ci.yml
Comment (cdeln) [reply .github/workflows/ci.yml:53]: I made an issue for it: https://github.com/cdeln/cpp_enum_set/issues/4

File: Makefile
Comment (cdeln) [reply Makefile:73]: I installed vcpkg in docker and can now build and test locally on my machine. I will push updated docker image in another commit.

File: BUILDING.md
Comment (cdeln) [reply BUILDING.md:71]: I will update this in a separate commit.

File: test/CMakeLists.txt
Comment (cdeln) [reply test/CMakeLists.txt:51]: Cool, I can reproduce everything locally on my machine now, this is ready for merge.

File: BUILDING.md
Comment (cdeln) [reply BUILDING.md:24]: Hmm it works through docker, must be my local setup that is just too old.

**Pull request merged**
